File: vbnet/GildedRoseKata/GildedRose.vb
Imports System.Collections.Generic

Public Class GildedRose
    Private Items As IList(Of Item)

    Public Sub New(items As IList(Of Item))
        Me.Items = items
    End Sub

    Public Sub UpdateQuality()
        For i = 0 To Items.Count - 1

            If Items(i).Name <> "Aged Brie" AndAlso Items(i).Name <> "Backstage passes to a TAFKAL80ETC concert" Then

                If Items(i).Quality > 0 Then

                    If Items(i).Name <> "Sulfuras, Hand of Ragnaros" Then
                        Items(i).Quality = Items(i).Quality - 1
                    End If
                End If
            Else

                If Items(i).Quality < 50 Then
                    Items(i).Quality = Items(i).Quality + 1

                    If Items(i).Name = "Backstage passes to a TAFKAL80ETC concert" Then

                        If Items(i).SellIn < 11 Then

                            If Items(i).Quality < 50 Then
                                Items(i).Quality = Items(i).Quality + 1
                            End If
                        End If

                        If Items(i).SellIn < 6 Then

                            If Items(i).Quality < 50 Then
                                Items(i).Quality = Items(i).Quality + 1
                            End If
                        End If
                    End If
                End If
            End If

            If Items(i).Name <> "Sulfuras, Hand of Ragnaros" Then
                Items(i).SellIn = Items(i).SellIn - 1
            End If

            If Items(i).SellIn < 0 Then

                If Items(i).Name <> "Aged Brie" Then

                    If Items(i).Name <> "Backstage passes to a TAFKAL80ETC concert" Then

                        If Items(i).Quality > 0 Then

                            If Items(i).Name <> "Sulfuras, Hand of Ragnaros" Then
                                Items(i).Quality = Items(i).Quality - 1
                            End If
                        End If
                    Else
                        Items(i).Quality = Items(i).Quality - Items(i).Quality
                    End If
                Else

                    If Items(i).Quality < 50 Then
                        Items(i).Quality = Items(i).Quality + 1
                    End If
                End If
            End If
        Next
    End Sub
End Class

File: csharp.NUnit/GildedRose/GildedRose.cs
using System.Collections.Generic;

namespace GildedRoseKata;

public class GildedRose
{
    IList<Item> Items;

    public GildedRose(IList<Item> Items)
    {
        this.Items = Items;
    }

    public void UpdateQuality()
    {
        foreach (var item in Items)
        {
            if (item.Name == "Sulfuras, Hand of Ragnaros")
                continue;

            item.SellIn--;

            switch (item.Name)
            {
                case "Aged Brie":
                    IncreaseQuality(item);
                    if (item.SellIn < 0)
                        IncreaseQuality(item);
                    break;

                case "Backstage passes to a TAFKAL80ETC concert":
                    if (item.SellIn < 0)
                    {
                        item.Quality = 0;
                    }
                    else
                    {
                        IncreaseQuality(item);
                        if (item.SellIn < 10) IncreaseQuality(item);
                        if (item.SellIn < 5) IncreaseQuality(item);
                    }
                    break;

                default:
                    DecreaseQuality(item);
                    if (item.SellIn < 0)
                        DecreaseQuality(item);
                    break;
            }
        }
    }

    private void IncreaseQuality(Item item)
    {
        if (item.Quality < 50) item.Quality++;
    }

    private void DecreaseQuality(Item item)
    {
        if (item.Quality > 0) item.Quality--;
    }
}